// nodetv-trakt/index.js
var crypto	= require('crypto'),
	extend	= require('xtend'),
	log4js	= require('log4js'),
	request	= require('request');

log4js.configure({
	appenders: [{
		type: 'console'
	}],
	replaceConsole: true
});
var logger = log4js.getLogger('nodetv-trakt');


var Trakt = function(settings){
	API.init(settings);
	return API;
}

var API = {
	init: function(obj){
		this.settings = extend(this.settings, obj);
		return this;
	},
	settings: {
		enabled: true,
		hashed: true,
		username: null,
		password: null,
		apikey: null
	},
	baseurl: 'http://api.trakt.tv/',
	
	/* Utility Methods */
	password: function() {
		try {
			if (!API.settings.hashed) return require('crypto').createHash('sha1').update(API.settings.password).digest('hex');
			return API.settings.password;
		} catch(e) {
			logger.error(e.message);
		}
	},
	
	get: function(url, data, callback) {
		if (typeof(data) == 'object') {
			var array = [];
			for (var i in data) array.push(data[i]);
			data = array;
		}
		var path = (data) ? '/' + data.join('/') : '';
		request({
			uri: API.baseurl+url+'/'+API.settings.apikey + path,
			method: 'GET'
		}, function(error, req, response){
			if (error) {
				logger.error(error);
				return;
			}
			try {
				if (typeof(response) != 'object') response = JSON.parse(response);
				if (response.error) return logger.error(response.error);
				if (typeof(callback) == 'function') callback(null, response);
			} catch(e) {
				logger.error(url+': '+e.message);
			}
		}).auth(API.settings.username, API.password());
	},
	
	post: function(url, data, callback) {
		request({
			uri: API.baseurl+url+'/'+API.settings.apikey,
			json: data,
			method: 'POST'
		}, function(error, req, response){
			if (error) {
				logger.error(url+': '+error);
				return;
			}
			try {
				if (typeof(response) != 'object') response = JSON.parse(response);
				if (response.error) return logger.error(response.error);
				if (typeof(callback) == 'function') callback(null, response);
			} catch(e) {
				logger.error(url+': '+e.message);
			}
		}).auth(API.settings.username, API.password());
	},
	
	search: function(type, query, callback){
		request({
			uri: 'http://api.trakt.tv/search/'+type+'.json/'+API.settings.apikey + '?limit=20&query=' + query,
			method: 'GET'
		}, function(error, req, response){
			if (error) {
				logger.error(error);
				return;
			}
			try {
				if (typeof(response) != 'object') response = JSON.parse(response);
			//	var error = (response.status == 'success') ? false : true;
				if (typeof(callback) == 'function') callback(null, response);
			} catch(e) {
				logger.error('search/'+type+': '+e.message);
			}
		}).auth(API.settings.username, API.password());
	},
	
	/*********************************/
	
	account: {
		test: function(callback){
			var payload = {};
			API.post('account/test', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		}
	},
	
	/*********************************/
	
	calendar: {
		shows: function(callback){
			var payload = {
				username: API.settings.username
			};
			API.get('user/calendar/shows.json', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		}
	},
	
	/*********************************/
	
	movie: {
		library: function(imdb, callback){
			var payload = {
				imdb_id: imdb
			};
			API.post('movie/library', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		seen: function(imdb, callback){
			var payload = {
				imdb_id: imdb
			};
			API.post('movie/seen', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		summary: function(imdb, callback){
			var payload = {
				imdb_id: imdb
			};
			API.get('movie/summary.json', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		unlibrary: function(imdb, callback){
			var payload = {
				imdb_id: imdb
			};
			API.post('movie/unlibrary', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		unseen: function(imdb, callback){
			var payload = {
				imdb_id: imdb
			};
			API.post('movie/unseen', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		}
	},
	
	/*********************************/
	
	movies: {
		trending : function(callback){
			
		},
		updated: function(callback){
			
		}
	},
	
	/*********************************/
	
	network: {
		follow: function(user, callback){
			var payload = {
				user: user
			};
			API.post('network/follow', payload, callback);
		}
	},
	
	/*********************************/
	
	recommendations: {
		shows: function(callback){
			var payload = {
				hide_collected: true,
				hide_watchlisted: true	
			};
			API.post('recommendations/shows', payload, callback);
		}
	},
	
	server: {
		time: function(){
			API.get('server/time.json', {}, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		}
	},
	
	/*********************************/
	
	show: {
		checkin: function(){
			var payload = {
				tvdb_id: null,
				season: null,
				episode: null
			}
			API.post('show/checkin', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		
		episode: {
			library: function(tvdb, list, callback){
				// Add episode to library
				/* list = {season: 1, episode: 1} */
				var payload = {
					tvdb_id: tvdb,
					episodes: list
				};
				API.post('show/episode/library', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			},
			seen: function(tvdb, season, episode, callback){
				// Mark episode as watched
				var payload = {
					tvdb_id: tvdb,
					episodes: [{
						season: season,
						episode: episode
					}]
				};
				API.post('show/episode/seen', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			},
			summary: function(tvdb, season, episode, callback){
				var payload = {
					tvdb_id: tvdb,
					season: season,
					episode: episode
				};
				API.get('show/episode/summary.json', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			},
			unseen: function(tvdb, season, episode, callback){
				var payload = {
					tvdb_id: tvdb,
					episodes: [{
						season: season,
						episode: episode
						
					}]
				};
				API.post('show/episode/unseen', payload, function(error, json){	
					if (typeof(callback) == 'function') callback(error, json);
				});
			}

		},
		
		library: function(tvdb, callback){
			var payload = {
				tvdb_id: tvdb
			};
			API.post('show/library', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		
		season: {
			info: function(tvdb, season, callback){
				var payload = [tvdb, season];
				API.get('show/season.json', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			},
			library: function(tvdb, season, callback){
				var payload = {
					tvdb_id: tvdb,
					season: season
				};
				API.post('show/season/library', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			},
			seen: function(tvdb, season, callback){
				var payload = {
					tvdb_id: tvdb,
					season: season
				};
				API.post('show/season/seen', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			}
		},
		seasons: function(tvdb, callback){
			var payload = {
				tvdb_id: tvdb
			};
			API.get('show/seasons.json', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		seen: function(tvdb, callback){
			var payload = {
				tvdb_id: tvdb
			};
			API.post('show/seen', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		summary: function(tvdb, callback){
			var payload = {
				tvdb_id: tvdb
			};
			API.get('show/summary.json', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		unlibrary: function(tvdb, callback){
			var payload = {
				tvdb_id: tvdb
			};
			API.post('show/unlibrary', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			});
		},
		watchlist: function(shows, callback){
			var payload = {
				shows: []
			};
			if (typeof(shows) == 'object') {
				shows.forEach(function(k,v){
					payload.shows.push({
						tvdb_id: k
					});
				});
			} else {
				payload.shows.push(shows)
			}
			API.post('show/watchlist', payload, function(error, json){
				if (typeof(callback) == 'function') callback(error, json);
			})
		}
	},
	
	/*********************************/
	
	user: {
		library: {
			movies: {
				all: function(callback){
					var payload = {
						username: API.settings.username,
						extended: true
					};
					API.get('user/library/movies/all.json', payload, function(error, json){
						if (typeof(callback) == 'function') callback(error, json);
					})
				}
			},
			shows: {
				all: function(callback){
					var payload = {
						username: API.settings.username
					};
					API.get('user/library/shows/all.json', payload, function(error, json){
						if (typeof(callback) == 'function') callback(error, json);
					});
				}
			}
		},
		progress: {
			collected: function(callback){
				var payload = {
					username: API.settings.username
				};
				API.get('user/progress/collected.json', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			},
			watched: function(tvdb, callback){
				var payload = {
					username: API.settings.username,
					tvdb_id: tvdb
				};
				API.get('user/progress/watched.json', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			}
		},
		watchlist: {
			movies: function(callback){
				var payload = {
					username: API.settings.username
				};
				API.get('user/watchlist/movies.json', payload, function(error, json){
					if (typeof(callback) == 'function') callback(error, json);
				});
			}
		}
	}
};

module.exports = exports = Trakt;
